### ISSUE TITLE
Might be nice to always call `exporter.finish_exporting`

### ISSUE THREAD

**Issue by romanchyla** [scrapy/scrapy#5537]
### Description

When an exception happens inside Exporter (for whatever reason) - itemcount never gets increased (that's OK). But then during closing, the code doesn't get chance to act

This is the place in the code:

https://github.com/scrapy/scrapy/blob/master/scrapy/extensions/feedexport.py#L357

Perhaps the lines 357 and 353 should have been swapped?

### Additional context

I'm using `slot.storage` to open a database; and an exporter to start a session -- store() gets called (to close connection to the database) but session didn't have chance to commit -- yes, there is nothing to commit, but still - given the semantics (also visible from the code -- when `file` is supposed to be a file descriptor it might be worth consideration to always call `finish_exporting` to give the file descriptor chance to react before it is closed by the parent object)

necessary caveat: I might be mis-using the API and it is a minor detail, so please forgive if this looks silly (and feature not a bug)

**Comment by wRAR** [comment]
Without looking at it for enough time, it looks like we should call `finish_exporting` if we called `start_exporting` before (not always, though), if we want to support various protocols and formats. The example I'm usually using when thinking about these things is an XML file that needs a closing tag if there was an opening one.

**Comment by HatimZ** [comment]
Can I solve this issue. Looks like just need to add the slot.finish_exporting() before the if condition in the def _close_slot(self, slot, spider):

**Comment by wRAR** [comment]
@HatimZ probably, but you need to make sure this code isn't called if `slot.start_exporting()` wasn't called (by reading the code, or maybe by making some test case). You'll also need to add some new test to test this behavior.

**Comment by HatimZ** [comment]
@wRAR .Okay. Are there any examples/resources for writing test cases?

**Comment by wRAR** [comment]
If you mean a test case for making sure it's ohly called when start_exporting() was called it would just be a test spider with some exporter enabled and some different edge cases like zero items, maybe an exception happening at some early point etc.

**Comment by HatimZ** [comment]
I checked the code to make sure that `slot.finish_exporting()`  isn't called if `slot.start_exporting()` wasn't called  and I have come to the conclusion that indeed its not happening anywhere in the file . 
Even if `slot.finish_exporting()` was called without the start being called, it wouldn't do anything because it checks for the flag/boolean set by start. Hence even if it gets called , it wont do any harm. Am I right?
So Can I proceed to do what I suggested here ? 

> Can I solve this issue. Looks like just need to add the 'slot.finish_exporting()' before the if condition in the 'def _close_slot(self, slot, spider)'
